src/tree/gpu_hist/evaluate_splits.cu

```diff
@@ -8,316 +8,269 @@
-      EvaluateSplitsKernel<kBlockThreads, GradientSumT>, left, right, evaluator,
-      dh::ToSpan(feature_best_splits));
+__device__ ScanElem<GradientSumT>
+ScanOp<GradientSumT>::operator() (ScanElem<GradientSumT> lhs, ScanElem<GradientSumT> rhs) {
+  ScanElem<GradientSumT> ret;
+  if (lhs.findex != rhs.findex || lhs.node_idx != rhs.node_idx || lhs.forward != rhs.forward) {
+    // Segmented Scan
+    return rhs;
+  }
+  if (((lhs.node_idx == 0) &&
+      (left.feature_set.size() != left.feature_segments.size()) &&
+      !thrust::binary_search(thrust::seq, left.feature_set.begin(), left.feature_set.end(),
+                             lhs.findex)) ||
+      ((lhs.node_idx == 1) &&
+      (right.feature_set.size() != right.feature_segments.size()) &&
+      !thrust::binary_search(thrust::seq, right.feature_set.begin(), right.feature_set.end(),
+                             lhs.findex))) {
+    // Column sampling
+    return rhs;
+  }
 
-  // Reduce to get best candidate for left and right child over all features
-  auto reduce_offset =
-      dh::MakeTransformIterator<size_t>(thrust::make_counting_iterator(0llu),
-                                        [=] __device__(size_t idx) -> size_t {
-                                          if (idx == 0) {
-                                            return 0;
-                                          }
-                                          if (idx == 1) {
-                                            return left.feature_set.size();
-                                          }
-                                          if (idx == 2) {
-                                            return combined_num_features;
-                                          }
-                                          return 0;
-                                        });
-  size_t temp_storage_bytes = 0;
-  auto num_segments = out_splits.size();
-  cub::DeviceSegmentedReduce::Sum(nullptr, temp_storage_bytes,
-                                  feature_best_splits.data(), out_splits.data(),
-                                  num_segments, reduce_offset, reduce_offset + 1);
-  dh::TemporaryArray<int8_t> temp(temp_storage_bytes);
-  cub::DeviceSegmentedReduce::Sum(temp.data().get(), temp_storage_bytes,
-                                  feature_best_splits.data(), out_splits.data(),
-                                  num_segments, reduce_offset, reduce_offset + 1);
+  ret = rhs;
+  ret.partial_sum = lhs.partial_sum + rhs.partial_sum;
+  return ret;
 }
 
 template <typename GradientSumT>
-void EvaluateSingleSplit(common::Span<DeviceSplitCandidate> out_split,
-                         TreeEvaluator::SplitEvaluator<GPUTrainingParam> evaluator,
-                         EvaluateSplitInputs<GradientSumT> input) {
-  EvaluateSplits(out_split, evaluator, input, {});
+ReduceElem<GradientSumT>
+__device__ ReduceValueOp<GradientSumT>::operator() (ScanElem<GradientSumT> e) {
+  ReduceElem<GradientSumT> ret;
```

Comment:
Can readability be improved here?

Reply:
Any suggestions? All this function does is to initialize the members of `ReduceElem`. Do you find the computation of `left_sum` and `right_sum` confusing?